applications/FluidDynamicsApplication/custom_processes/distance_smoothing_process.cpp

```diff
@@ -0,0 +1,200 @@
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Mohammad R. Hashemi
+//
+
+// Project includes
+#include "includes/define.h"
+#include "containers/model.h"
+#include "includes/checks.h"
+#include "spaces/ublas_space.h"
+#include "modeler/connectivity_preserve_modeler.h"
+#include "utilities/variable_utils.h"
+#include "includes/global_pointer_variables.h"
+#include "utilities/parallel_utilities.h"
+#include "factories/linear_solver_factory.h"
+#include "linear_solvers/linear_solver.h"
+#include "solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h"
+#include "solving_strategies/schemes/residualbased_incrementalupdate_static_scheme.h"
+#include "solving_strategies/strategies/residualbased_linear_strategy.h"
+
+// Application includes
+#include "fluid_dynamics_application_variables.h"
+#include "distance_smoothing_process.h"
+#include "custom_elements/distance_smoothing_element.h"
+
+
+namespace Kratos
+{
+
+/* Public functions *******************************************************/
+template< unsigned int TDim, class TSparseSpace, class TDenseSpace, class TLinearSolver>
+DistanceSmoothingProcess<TDim, TSparseSpace, TDenseSpace, TLinearSolver>::DistanceSmoothingProcess(
+    ModelPart& rModelPart,
+    typename TLinearSolver::Pointer plinear_solver)
+    : Process(),
+      mrModelPart(rModelPart)
+{
+    // Generate an auxilary model part and populate it by elements of type DistanceSmoothingElement
+    CreateAuxModelPart();
+
+    auto p_builder_solver = Kratos::make_shared<ResidualBasedBlockBuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver> >(plinear_solver);
+
+    CreateSolutionStrategy(plinear_solver, p_builder_solver);
+}
+
+template< unsigned int TDim, class TSparseSpace, class TDenseSpace, class TLinearSolver>
+DistanceSmoothingProcess<TDim, TSparseSpace, TDenseSpace, TLinearSolver>::DistanceSmoothingProcess(
+    ModelPart& rModelPart,
+    Parameters Parameters)
+    : DistanceSmoothingProcess(
+        rModelPart,
+        LinearSolverFactory<TSparseSpace, TDenseSpace>().Create(Parameters["linear_solver_settings"])
+    ){}
+
+template< unsigned int TDim, class TSparseSpace, class TDenseSpace, class TLinearSolver>
+DistanceSmoothingProcess<TDim, TSparseSpace, TDenseSpace, TLinearSolver>::DistanceSmoothingProcess(
+    Model &rModel,
+    Parameters Parameters)
+    : DistanceSmoothingProcess(
+        rModel.GetModelPart(Parameters["model_part_name"].GetString()),
+        Parameters
+    ){}
+
+template< unsigned int TDim, class TSparseSpace, class TDenseSpace, class TLinearSolver>
+void DistanceSmoothingProcess<TDim, TSparseSpace, TDenseSpace, TLinearSolver>::CreateAuxModelPart()
+{
+    Model& current_model = mrModelPart.GetModel();
+    const std::string smoothing_model_part_name = mrModelPart.Name()+"_Smoothing";
```

Comment:
```suggestion
    const std::string smoothing_model_part_name = mrModelPart.FullName()+"_Smoothing";
```

just in case some submodelpart has the same name as a root modelpart

Reply:
Done.